README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# flexwinn

<!-- badges: start -->
<!-- badges: end -->

The goal of flexwinn is to correct a data set of metabolites that has
been distorted during the data gathering. There are two different
function in it.The first one (flexwinn) detects automatically through
dynamic programming the change points in time series and then
residualize, detrend and normalize the time series of metabolites. The
second one uses the plates information to correct to correct the
distorted time series. Both correct the time series in a very similar
way but on estimates the change points and one consider change points
coincide with plates boundaries.

## Installation

You can install the development version of flexwinn from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("antoinejeanrenaud/flexwinn")
```

## Example

This is a basic example which shows how to use the flexwinn function.
This example allows to correct for the two metabolites in the data set
met.df.

``` r
library(flexwinn)
# Load the data of metabolites
data(met.df)
# Use flexwinn to correct the metabolites
corrected<-flexwinn(as.data.frame(met.df),graph=FALSE)
```

There is a parameter “graph” which allow to plot the change points found
in the data sequence and show how it is detrended. When set to TRUE it
will produce a plot for the first metabolite of the data set to be
corrected.

``` r
library(flexwinn)
# Load the data of metabolites
data(met.df)
# Use flexwinn to correct the metabolites
met1<-as.data.frame(met.df$met1)
corrected<-flexwinn(met1,graph=TRUE)
```

<img src="man/figures/README-example_2-1.png" width="100%" /><img src="man/figures/README-example_2-2.png" width="100%" />

Here we have a metabolite that is not distorted, so the function
flexwinn is not really correcting it. The algorithm has found no change
point which is correct.

``` r
library(flexwinn)
# Load the data of metabolites
data(met.df)
# Use flexwinn to correct the metabolites
met2<-as.data.frame(met.df$met2)
corrected<-flexwinn(met2,graph=TRUE)
```

<img src="man/figures/README-example_3-1.png" width="100%" /><img src="man/figures/README-example_3-2.png" width="100%" />

This time the signal was distorted and we see the change points
estimated and the correction applied. Of course, the algorithm is
probably not recovering each change point perfectly, but in the end the
correlation with the original non distorted time series is the most
important. We can also have a look on how to use the winn function. The
only difference in use is that you also have to specify the places of
change in plate in an ordered vector.

``` r
library(flexwinn)
# Load the data of metabolites
data(met.df)
# Use winn to correct the metabolites
corrected<-winn(as.data.frame(met.df),c(53,105,177,274,330,415,519,605,688,771),graph=FALSE)
```

There is a parameter “graph” which allow to plot the plates in the data
sequence and show how it is detrended. When set to TRUE it will produce
a plot for the first metabolite of the data set to be corrected.

``` r
library(flexwinn)
# Load the data of metabolites
data(met.df)
# Use winn to correct the metabolites
met1<-as.data.frame(met.df$met1)
corrected<-winn(met1,c(53,105,177,274,330,415,519,605,688,771),graph=TRUE)
```

<img src="man/figures/README-example_5-1.png" width="100%" /><img src="man/figures/README-example_5-2.png" width="100%" />
Here we have a metabolite that is not distorted, so the function winn is
not really correcting it. Let’s now have a look at a simnulated
metabolite with batch effects and drifts.

``` r
library(flexwinn)
# Load the data of metabolites
data(met.df)
# Use winn to correct the metabolites
met2<-as.data.frame(met.df$met2)
corrected<-winn(met2,c(53,105,177,274,330,415,519,605,688,771),graph=TRUE)
```

<img src="man/figures/README-example_6-1.png" width="100%" /><img src="man/figures/README-example_6-2.png" width="100%" />
The plates are displayed, the correction applied is displayed. We have
set the plates here to coincide perfectly with the plates boundaries.
The time series displayed here is different than with the flexwinn
function because there is an additional step before detrending.


NEWS.md
# flexwinn 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.
* First submission after creating the package


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(flexwinn)

test_check("flexwinn")


tests/testthat/test-winn.R
test_that("winn input", {
  expect_error(winn(NULL,100))
  met1<-c(rnorm(1000),NA)
  met1<-as.data.frame(met1)
  expect_error(winn(met1,c(100,300,600,800)))
  met1<-NA
  met1<-as.data.frame(met1)
  expect_error(winn(met1))
})

test_that("winn works", {
  set.seed(333)
  met1<-c(rnorm(200),rnorm(200,sd=2)+5+seq(0,3,length.out=200))
  met1<-as.data.frame(met1)
  result<-winn(met1,end.plates=200,graph = TRUE)
  expect_type(result,"list")

  met1<-c(rnorm(200),rnorm(200)+5+seq(0,3,length.out=200))
  met1<-as.data.frame(met1)
  result<-winn(met1,end.plates=200,graph = FALSE)
  expect_type(result,"list")

  met1<-c(rep(0,200),rep(5,200))
  met1<-as.data.frame(met1)
  result<-winn(met1,end.plates=200,graph = FALSE)
  expect_type(result,"list")

  met1<-c(rep(0,200),rnorm(200,mean=5))
  met1<-as.data.frame(met1)
  result<-winn(met1,end.plates=200,graph = FALSE)
  expect_type(result,"list")
})


tests/testthat/test-flexwinn.R
test_that("fkPELT works", {
  expect_identical(fkPELT(rep(10,1000),NULL), numeric(0))
  expect_equal(fkPELT(c(rep(10,1000),rep(20,1000)),NULL), 1000)
  expect_identical(fkPELT(1:1000,NULL),numeric(0))
  expect_error(fkPELT(NULL,NULL))
})
test_that("flexwinn input", {
  expect_error(flexwinn(NULL))
  met1<-c(rnorm(1000),NA)
  met1<-as.data.frame(met1)
  expect_warning(flexwinn(met1))
  met1<-NA
  met1<-as.data.frame(met1)
  expect_error(flexwinn(met1))
})

  test_that("flexwinn works", {
  met1<-rep(15,1000)
  met1<-as.data.frame(met1)
  expect_identical(flexwinn(met1),met1-15)

  set.seed(333)
  met1<-c(rnorm(200),rnorm(200,sd=2)+5+seq(0,3,length.out=200))
  met1<-as.data.frame(met1)
  result<-flexwinn(met1,graph = TRUE)
  expect_type(result,"list")

  met1<-c(rnorm(200),rnorm(200)+5+seq(0,3,length.out=200))
  met1<-as.data.frame(met1)
  result<-flexwinn(met1,graph = FALSE)
  expect_type(result,"list")

  met1<-c(rep(0,200),rep(5,200))
  met1<-as.data.frame(met1)
  result<-flexwinn(met1,graph = FALSE)
  expect_type(result,"list")

  met1<-c(rep(0,200),rnorm(200,mean=5))
  met1<-as.data.frame(met1)
  result<-flexwinn(met1,graph = FALSE)
  expect_type(result,"list")
})




R/winn.R
#' @title winn
#' @description A function for metabolite correction.
#' This function corrects batch effects and drifts
#' with the plate information by detrending and normalizing when needed.
#' @param data input dataset of metabolites as data frame.
#' @param graph TRUE or FALSE. If TRUE will display the graph of uncorrected
#' data with plates and correction
#' @param end.plates vector containing the position of the changes in plates
#' in terms of run order. The vector must be ordered.
#' @return the corrected dataframe of metabolite(s).
#' @import splines
#' @import dplyr
#' @import lawstat
#' @import mgcv
#' @importFrom graphics abline legend lines
#' @importFrom stats na.omit Box.test
#' @export
#' @examples
#' met1<-rnorm(200)
#' winn(as.data.frame(met1),end.plates=100)

winn<-function(data,end.plates,graph=FALSE){
  if (!is.data.frame(data)){
    stop("data must be a data frame") # We ensure the input is of type
    # data.frame
  }
  if (is.null(end.plates)){
    stop("end.plates cannot be empty") # We ensure there is at least two
    # plates
  }
  NA_present<-FALSE
  for (k.na in 1:dim(data)[2]) {
    if (sum(is.na(data[,k.na]))>0){
      NA_present<-TRUE
    }
  }
  if (NA_present){
      stop("There is(are) NA value(s) in the data frame, please remove Na's")

  }
  corrected<-data
  for (k in 1:dim(data)[2]){

    changepoints<-end.plates
    colname<-colnames(data[k])
    if (is.null(colname)){colnames(data[k])<-paste("met_",k,sep="")}
    changes<-length(changepoints)
    datanew<-data[[k]]-mean(data[[k]])
    #####
    #First normalization
    #This normalization is for the next anova test
    data.norm<-normalize.var(datanew,changepoints)
    ######
    #Residualizing if needed
    data.resid<-residualize(data.norm,changepoints)
    #############
    #Detrending
    tau<-c(0,changepoints,length(data.resid))
    pred<-NULL #create a variable to store correction
    for (c in 1:(changes+1)){
      subset<-data.resid[(tau[c]+1):tau[c+1]]
      size<-length(subset)
      if (size>=5){
        pvalue<-Box.test(subset,lag=floor(length(subset)/2),
                         type="Ljung-Box")$p.value
        # Test if the segment has autocorrelation
        if (!is.na(pvalue) & pvalue<0.01){
          dfw<-dfwhitenoise(subset)#Find the best degree of freedom
          x<-1:length(subset)
          spline<-mgcv::gam(subset~s(x,bs="cr",k=dfw,fx=TRUE))
          pred<-c(pred,spline$fitted.values)#correction that will be applied
        }else{
          pred<-c(pred,rep(0,size)) #store the correction
        }

      }else{
        pred<-c(pred,rep(0,size))
      }
    }
    if (k==1){
      if (graph==TRUE){
        ## here is the code to plot visualizatio of how the algorithm works
        ## We display change points together with uncorrected signal and
        ## the correction that will be applied
        concentration<-data.resid
        plot(concentration,col="darkgray",
             xlab="reading sequence",ylab="residualized/normalized signal",
             main="residualized/normalized signal + correction and plates",
             type="l")
        for (l in changepoints) {
          abline(v=l,col="red",lty=2)
        }
        lines(1:length(data.resid),pred,col="red")
        legend(
          "bottomleft",
          lty = c(1, 1),
          col = c("darkgrey", "red"),
          legend = c("residualized/normalized", "Correction"),
          cex=0.5
        )
      }}
    datanew<-data.resid-pred
    #######
    #normalizing by segment again
    data.norm.2<-normalize.var(datanew,changepoints)
    ######
    #Residualizing if needed
    data.resid.2<-residualize(data.norm.2,changepoints)
    #######
    #return result
    corrected[[k]]<-data.resid.2

  }

  return(corrected)

}

#'@noRd
#'@importFrom stats aov na.exclude
residualize<-function(met.data,endpoints){
  resid<-FALSE
  met.data<-changetometa(as.data.frame(met.data),endpoints)
  colnames(met.data)<-"orig"
  plates <-
    unique(unlist(lapply(strsplit(
      rownames(met.data), "_"
    ), function(x) {
      return(x[2])
    })))
    if (length(plates)<=1){
      a.pval<-1
    }else{
      met.data[["group"]] <-
        as.factor(paste("plate", "_",
                        unlist(lapply(strsplit(rownames(met.data), "_"),
                          function(x) {return(x[2])})), sep = ""))

      # ANOVA test
      a <- aov(as.formula(paste("orig", "~ group", sep = "")), data = met.data)
      a.pval <- summary(a)[[1]][, "Pr(>F)"][1]
    }
      if (a.pval<0.05){
        resid<-TRUE
      }


    if (resid==TRUE){
      for (i in 1:(length(plates) - 1)) {
        on.plate <- paste("in.", i, sep = "")
        met.data[[on.plate]] <- 0
        met.data[grep(paste("plate", "_", plates[i], "_", sep = ""),
                    rownames(met.data)),][[on.plate]] <- 1
      }

      # Formula
      form <-
        paste("orig", " ~ ", paste(paste(
          "as.factor(in.", 1:(length(plates) - 1), ")", sep = ""
        ), collapse = " + "))

      # Run regression and residualize if no errors
      ret <- tryCatch({
        reg <- lm(as.formula(form), data = met.data, na.action = na.exclude)

# Calling "resid" on the regression object created with the na.action=na.exclude
# option takes care of the missing values. "resid(reg)" return NA if the
# observation is also NA (remember that reg$resid will only return the
# residuals for the non-missing values)
        list(res = resid(reg), status = 0)
      },
      error = function(e) {
        print(paste("Error processing", "orig"))
        print(e)

        # Residualization failed, return a vector of "res.failed"
        return(list(res = rep("res.failed", dim(met.data)[1]), status = -1))
      }) # End of try-catch






      if (ret$status == 0) {
        data.resid<-ret$res

      }else{
        data.resid<-met.data[["orig"]]
      }
    }else{
      data.resid<-met.data[["orig"]]
    }


  return(data.resid)
}


R/changomics.R
#' @title changomics
#' @description A function for metabolite correction.
#' This function first estimate the change points
#'  (jumps) and then correct each different segment
#'  by detrending and normalizing when needed.
#' @param data input dataset of metabolites as data frame.
#' @param graph TRUE or FALSE. If TRUE will display the graph of uncorrected
#' data with estimated change points and correction
#' @return the corrected dataframe of metabolite(s).
#' @import splines
#' @import dplyr
#' @import lawstat
#' @importFrom graphics abline legend lines
#' @importFrom stats na.omit Box.test
#' @export
#' @examples
#' met1<-rnorm(200)
#' changomics(as.data.frame(met1))

changomics<-function(data,graph=FALSE){
  if (!is.data.frame(data)){
    stop("data must be a data frame") # We ensure the input is of type
                                      # data.frame
  }
  NA_present<-FALSE
  for (k.na in 1:dim(data)[2]) {
    if (sum(is.na(data[,k.na]))){
      NA_present<-TRUE
    }
  }
  if (NA_present){
    data<-na.omit(data) #remove NAs
    if (nrow(data)>0){
    warning("There is(are) NA value(s) in the data frame, na.omit()
    removed the rows containing NA(s).
            recommended-> remove Na's yourself to not remove a full row")
      }

  }
  if (nrow(data)==0){
    stop("All rows have been removed because there were Na's in each row:
    remove yourself Na's and run algo
         for one metabolites at a time") #If all values have been removed
    }
  corrected<-data
  for (k in 1:dim(data)[2]){
    knots<-seq(1,length(data[[k]]),length.out=floor(length(data[[k]])/60)+2)
    #Select number of knots ~1 knot per 60 observations
    knots<-knots[-c(1,length(knots))]
    knots<-knots+rep(0.5,length(knots))*(floor(knots)==knots)
    changepoints<-fkPELT(data[[k]],knots = knots) # find the change points
    colname<-colnames(data[k])
    if (is.null(colname)){colnames(data[k])<-paste("met_",k,sep="")}
    #############
    #Detrending
    changes<-length(changepoints)
    datanew<-data[[k]]
    tau<-c(0,changepoints,length(data[[k]]))
    pred<-NULL #create a variable to store correction
    for (c in 1:(changes+1)){
      subset<-datanew[(tau[c]+1):tau[c+1]]
      size<-length(subset)
      if (size>=5){
        pvalue<-Box.test(subset,lag=floor(length(subset)/2),
                         type="Ljung-Box")$p.value
        # Test if the segment has autocorrelation
        if (!is.na(pvalue) & pvalue<0.05){
        dfw<-dfwhitenoise(subset) #Find the best degree of freedom
        spline<-smooth.spline(subset,df=dfw)
        pred<-c(pred,spline$y)#correction that will be applied
      }else{
        pred<-c(pred,rep(mean(subset),size)) #store the correction
      }

      }else{
        pred<-c(pred,rep(mean(subset),size))
      }
    }
    if (k==1){
    if (graph==TRUE){
      ## here is the code to plot visualizatio of how the algorithm works
      ## We display change points together with uncorrected signal and
      ## the correction that will be applied
      concentration<-data[[k]]
      plot(concentration,col="darkgray",
           xlab="reading sequence",ylab="uncorrected signal",
           main="uncorrected signal + correction and change points",
           type="l")
      for (l in changepoints) {
        abline(v=l,col="red",lty=2)
      }
      lines(1:length(data[[k]]),pred,col="red")
      legend(
        "bottomleft",
        lty = c(1, 1),
        col = c("darkgrey", "red"),
        legend = c("Uncorrected", "Correction"),
        cex=0.5
      )
    }}
    datanew<-datanew-pred
    #######
    #normalizing
    datanew<-changetometa(as.data.frame(datanew),changepoints)
    ## Put in form that we can know which segment/plate the data is in
    colnames(datanew)<-"orig"
    if (changes>=1){
      ## if no change no normalization
      homo2<-homogen.var.2(datanew) # Test if the different segments
                                    # have homogeneity of variance
      if (!homo2){

        plates <-
          unique(unlist(lapply(strsplit(
            rownames(datanew), "_"
          ), function(x) {
            return(x[2])
          })))

        met.plates.sd <- c()

        for (j in plates) {

          grep.plates <-
            grep(paste("plate", "_", j, "_", sep = ""), row.names(datanew))
          sd.plate    <-  sd(datanew[grep.plates,], na.rm = TRUE)



          if(!is.na(sd.plate) & sd.plate == 0) {
            sd.plate <- NA
          }

          met.plates.sd <-
            c(met.plates.sd, rep(sd.plate, length(grep.plates)))
        }

        # Normalize
        datanew[["norm"]] <- datanew[["orig"]] / met.plates.sd
        if (sum(is.na(datanew[["norm"]]))>0){

          dataret<-datanew[["orig"]]

        }

        dataret<-datanew[["norm"]]
      }
    }

    dataret<-datanew[["orig"]]


  #######
  #return result
    corrected[[k]]<-dataret
  }

  return(corrected)

}
#########
##change data to be in the form with group/plate to then normalize
changetometa<-function(data,changepoints){
  names<-vector("character",length = length(data[[1]]))
  tau<-c(0,changepoints,length(data[[1]]))
  for (i in 1:(length(tau)-1)){
    for (p in (tau[i]+1):tau[i+1]){
      plate_number<-i
      order<-p

      names[p]<-paste("plate",plate_number,"order",order,sep = "_")
    }
  }
  colnam<-colnames(data)
  ret<-as.data.frame(data[[1]])
  colnames(ret)<-colnam
  row.names(ret)<-names
  return(ret)
}

#############
#'@noRd
#'@importFrom stats smooth.spline

## Find the best degree of freedom that maximize the p-value of
## autocorrelation test
dfwhitenoise<-function(data){
  t<-n_distinct(data)
  l<-min(c(t-1,floor(t*20/100)))
  value<-numeric(length = l-1)
  for (i in 2:l){
    spline<-smooth.spline(data,df=i)
    epsilon<-data-spline$y
    value[i-1]<-Box.test(epsilon,lag=15,type="Ljung-Box")$p.value
  }
  df<-which.max(value)+1
  return(df)
}
###########
#' @noRd
#' @importFrom stats sd shapiro.test fligner.test as.formula

## Test if the different segments in the data set have the same variance
homogen.var.2 <- function(met.dat)
{
  # Create a vector to store results of homogeneity variance test
  is.hom.var <- FALSE

    met.df <-  met.dat["orig"]

    if(sd(met.dat[,1], na.rm=TRUE) == 0){
      is.hom.var <- TRUE}
    else{
      met.df[["group"]] <- sapply(rownames(met.df),
                                  function(x) {
                                    strsplit(x, "_")[[1]][2]
                                  })

      # Shapiro test of normality
      x <- shapiro.test(met.df[, 1])

      # If normality  use Levine test for homogeneity of variance
      # Else us Fligner-Killeen test
      p.val <- c()
      if (x$p.value < 0.05) {
        # no normality : Fligner-Killeen
        flig.tst <-
          fligner.test(as.formula(paste("orig", " ~ group", sep = "")),
                       data = met.df)
        p.val <- flig.tst$p.value
      } else{
        # Change this once bartlett.test from "car" library is available
        lev.tst <- levene.test(met.df[, 1],
                               met.df[, "group"],
                               location = "median",
                               correction.method = "zero.correction")
        p.val <- lev.tst$p.value
      }
      if (p.val > 0.05) {
        is.hom.var<- TRUE
      }
    }


  return(is.hom.var)
}

########
#Cost of segment calculator with fixed knots
#' @noRd
#' @importFrom stats lm
fksplinecost<-function(data,knots,index1=1,index2=length(data)){
  size<-length(data)
  if (size==1){
    return(0)
  }
  #if size is too small, we do not fit a spline
  #we just use the mean as an estimate.
  if (size<5){
    sd<-sd(data)
    mu<-mean(data)
    neglog<-2*(sum((data-mu)^2/(2*sd^2))+size*log(sd*sqrt(2*pi)))
    # If sd is too small we return 0 cost
    if (sd<=1e-5){
      return(0)
    }else{
      return(neglog)
    }
  }
  cov<-index1:index2
  newknots<-knots[knots<index2& knots>index1]
  # We take only the knots that are inside both start index and end index
  spline<-lm(data~ns(cov,knots = newknots,intercept=TRUE))
  # We estimate the moving mean (drift) with a spline
  mu<-spline$fitted.values
  sd<-sd(data-mu) #from this estimated moving mean we can estimate the sd
  # If sd is too small we return 0 cost
  if (sd<=1e-5){
    return(0)
  }
  neglog<-2*(sum((data-mu)^2/(2*sd^2))+size*log(sd*sqrt(2*pi)))
  #log likelihood of the segment supposing it is close to a
  #normal distribution (usually the case with metabolomics data set)
  return(neglog)


}



########
#Estimating changepoints in time series (with drifts and jumps)
# with fixed knots predefined
#This algo is PELT algorithm but with a slight modification in the pruning and
#the cost of a segement as defined with the previous function "fksplinecost"
fkPELT<-function(data,knots){
  # Check if data is empty
  if (is.null(data)){stop("data cannot be NULL in fkPELT")}
  n<-length(data)
  f<-numeric(length = n+1)
  f[1]<- -3*log(300) # 3*log(300) is the penalty constant tha we use
                     #in the PELT algorithm. Inspired by BIC criterion but
                     #modified
  cp<-rep(list(numeric()),n+1)
  R<-rep(list(numeric()),n)
  R[[1]]<-0


  for (t in 1:n){
    m<-length(R[[t]])
    neglog<-numeric(m)
    for (r in 1:m){
      neglog[r]<-fksplinecost(data[(R[[t]][r]+1):t],
                              knots=knots,index1=R[[t]][r]+1,index2=t)
    }
    stat<-numeric(m)
    for (r in 1:m){
      # Calculate the global cost with each possible change points in R[[t]]
      tot<-f[R[[t]][r]+1]+3*log(300)+neglog[r]
      stat[r]<-tot

    }
    f[t+1]<-min(stat)
    t1<-R[[t]][which.min(stat)]
    cp[[t+1]]<-c(cp[[t1+1]],t1)
    for (r in 1:m){

      tot<-f[R[[t]][r]+1]+log(300)+neglog[r]
      #Here we have the pruning which is amplified by this "log(300)"
      #, this makes the method a lot faster but a little bit less accurate
      if (tot<=f[t+1]& t<n){

        R[[t+1]]<-c(R[[t+1]],R[[t]][r])
      }
    }
    if (t<n&t>39){
      R[[t+1]]<-c(R[[t+1]],t-19)}
  }

  cp[[n+1]]<-cp[[n+1]][-1]
  return(cp[[n+1]]) #Return the set of change points
}









R/flexwinn.R
#' @title flexwinn
#' @description A function for metabolite correction.
#' This function first estimate the change points
#'  (jumps) and then correct each different segment
#'  by detrending and normalizing when needed.
#' @param data input dataset of metabolites as data frame.
#' @param graph TRUE or FALSE. If TRUE will display the graph of uncorrected
#' data with estimated change points and correction and the corrected signal
#' on a second plot.
#' @return the corrected dataframe of metabolite(s).
#' @import splines
#' @import dplyr
#' @import lawstat
#' @import mgcv
#' @importFrom graphics abline legend lines
#' @importFrom stats na.omit Box.test
#' @export
#' @examples
#' met1<-rnorm(200)
#' flexwinn(as.data.frame(met1))

flexwinn<-function(data,graph=FALSE){
  if (!is.data.frame(data)){
    stop("data must be a data frame") # We ensure the input is of type
                                      # data.frame
  }
  NA_present<-FALSE
  for (k.na in 1:dim(data)[2]) {
    if (sum(is.na(data[,k.na]))>0){
      NA_present<-TRUE
    }
  }
  if (NA_present){
    data<-na.omit(data) #remove NAs
    if (nrow(data)>0){
    warning("There is(are) NA value(s) in the data frame, na.omit()
    removed the rows containing NA(s).
            recommended-> remove Na's yourself to not remove a full row")
      }

  }
  if (nrow(data)==0){
    stop("All rows have been removed because there were Na's in each row:
    remove yourself Na's and run algo
         for one metabolites at a time") #If all values have been removed
    }
  corrected<-data
  for (k in 1:dim(data)[2]){
    knots<-seq(1,length(data[[k]]),length.out=floor(length(data[[k]])/60)+2)
    #Select number of knots ~1 knot per 60 observations
    knots<-knots[-c(1,length(knots))]
    knots<-knots+rep(0.5,length(knots))*(floor(knots)==knots)
    #make sure the knot is not on an observation
    changepoints<-fkPELT(data[[k]],knots = knots) # find the change points
    colname<-colnames(data[k])
    #If there is no name give one
    if (is.null(colname)){colnames(data[k])<-paste("met_",k,sep="")}
    #############
    #Detrending
    changes<-length(changepoints)
    datanew<-data[[k]]
    tau<-c(0,changepoints,length(data[[k]]))
    pred<-NULL #create a variable to store correction
    for (c in 1:(changes+1)){
      #take each segment
      subset<-datanew[(tau[c]+1):tau[c+1]]
      size<-length(subset)
      if (size>=5){
        pvalue<-Box.test(subset,lag=floor(length(subset)/2),
                         type="Ljung-Box")$p.value
        # Test if the segment has autocorrelation
        if (!is.na(pvalue) & pvalue<0.01){
        dfw<-dfwhitenoise(subset) #Find the best degree of freedom
        x<-1:length(subset)
        spline<-mgcv::gam(subset~s(x,bs="cr",k=dfw,fx=TRUE))
        pred<-c(pred,spline$fitted.values)#correction that will be applied
      }else{
        pred<-c(pred,rep(mean(subset),size)) #store the correction
      }

      }else{
        pred<-c(pred,rep(mean(subset),size))
      }
    }
    if (k==1){
    if (graph==TRUE){
      ## here is the code to plot visualization of how the algorithm works
      ## We display change points together with uncorrected signal and
      ## the correction that will be applied
      concentration<-data[[k]]
      plot(concentration,col="darkgray",
           xlab="reading sequence",ylab="uncorrected signal",
           main="uncorrected signal + correction and change points",
           type="l")
      for (l in changepoints) {
        abline(v=l,col="red",lty=2)
      }
      lines(1:length(data[[k]]),pred,col="red")
      legend(
        "bottomleft",
        lty = c(1, 1),
        col = c("darkgrey", "red"),
        legend = c("Uncorrected", "Correction"),
        cex=0.5
      )
    }}
    datanew<-datanew-pred
    #######
    #normalizing by segment
    data.norm<-normalize.var(datanew,changepoints)
    #####
    #plot corrected
    if (k==1){
      if (graph==TRUE){

        ##We display the corrected signal for the first metabolite
        concentration<-data.norm
        plot(concentration,col="darkgray",
             xlab="reading sequence",ylab="corrected signal",
             main="corrected signal",
             type="l")

        legend(
          "bottomleft",
          lty = c(1, 1),
          col = c("darkgrey"),
          legend = c("corrected"),
          cex=0.5
        )
      }
      }
    #######
   #return result
    corrected[[k]]<-data.norm
  }

  return(corrected)

}
#########
##change data to be in the form with group/plate to then normalize
changetometa<-function(data,changepoints){
  names<-vector("character",length = length(data[[1]]))
  tau<-c(0,changepoints,length(data[[1]]))
  for (i in 1:(length(tau)-1)){
    for (p in (tau[i]+1):tau[i+1]){
      plate_number<-i
      order<-p

      names[p]<-paste("plate",plate_number,"order",order,sep = "_")
    }
  }
  colnam<-colnames(data)
  ret<-as.data.frame(data[[1]])
  colnames(ret)<-colnam
  row.names(ret)<-names
  return(ret)
}

#############
#'@noRd
#'@importFrom stats residuals


## Find the best degree of freedom that maximize the p-value of
## autocorrelation test o the residuals. For the spline fit we are using mgcv::gam
dfwhitenoise<-function(data){
  t<-n_distinct(data)
  l<-max(min(c(t-1,floor(t*10/100))),3)
  value<-numeric(length = l-2)
  #loop to test different fit
  for (i in 3:l){
    x<-1:length(data)
    spline<-mgcv::gam(data~s(x,bs="cr",k=i,fx=TRUE))
    epsilon<-residuals(spline)
    value[i-2]<-Box.test(epsilon,lag=min(20,length(epsilon)-1),type="Ljung-Box")$p.value
  }
  df<-which.max(value)+2
  return(df)
}
###########
#' @noRd
#' @importFrom stats sd shapiro.test fligner.test as.formula

## Test if the different segments in the data set have the same variance
homogen.var <- function(met.dat)
{
  # Set to FALSE first
  is.hom.var <- FALSE

    met.df <-  met.dat["orig"]
    if(sd(met.dat[,1], na.rm=TRUE) == 0){
      is.hom.var <- TRUE
      }else{
      met.df[["group"]] <- sapply(rownames(met.df),
                                  function(x) {
                                    strsplit(x, "_")[[1]][2]
                                  })

      # Shapiro test of normality
      x <- shapiro.test(met.df[, 1])

      # If normality  use Levine test for homogeneity of variance
      # Else us Fligner-Killeen test
      p.val <- c()
      if (x$p.value < 0.05) {
        # no normality : Fligner-Killeen
        flig.tst <-
          fligner.test(as.formula(paste("orig", " ~ group", sep = "")),
                       data = met.df)
        p.val <- flig.tst$p.value
      } else{
        # Change this once bartlett.test from "car" library is available
        lev.tst <- levene.test(met.df[, 1],
                               met.df[, "group"],
                               location = "median",
                               correction.method = "zero.correction")
        p.val <- lev.tst$p.value
      }
      if(!is.na(p.val)){
      if (p.val > 0.05) {
        is.hom.var<- TRUE
      }
      }
    }


  return(is.hom.var)
}

########
#Cost of segment calculator with fixed knots
#' @noRd
#' @importFrom stats lm
fksplinecost<-function(data,knots,index1=1,index2=length(data)){
  size<-length(data)
  if (size==1){
    return(0)
  }
  #if size is too small, we do not fit a spline
  #we just use the mean as an estimate.
  if (size<5){
    sd<-sd(data)
    mu<-mean(data)
    neglog<-2*(sum((data-mu)^2/(2*sd^2))+size*log(sd*sqrt(2*pi)))
    # If sd is too small we return 0 cost
    if (sd<=1e-5){
      return(0)
    }else{
      return(neglog)
    }
  }
  cov<-index1:index2
  newknots<-knots[knots<index2& knots>index1]
  # We take only the knots that are inside both start index and end index
  spline<-lm(data~ns(cov,knots = newknots,intercept=TRUE))
  # We estimate the moving mean (drift) with a spline
  mu<-spline$fitted.values
  sd<-sd(data-mu) #from this estimated moving mean we can estimate the sd
  # If sd is too small we return 0 cost
  if (sd<=1e-5){
    return(0)
  }
  neglog<-2*(sum((data-mu)^2/(2*sd^2))+size*log(sd*sqrt(2*pi)))
  #log likelihood of the segment supposing it is close to a
  #normal distribution (usually the case with metabolomics data set)
  return(neglog)


}



########
#Estimating changepoints in time series (with drifts and jumps)
# with fixed knots predefined
#This algo is PELT algorithm but with a slight modification in the pruning and
#the cost of a segement as defined with the previous function "fksplinecost"
fkPELT<-function(data,knots){
  # Check if data is empty
  if (is.null(data)){stop("data cannot be NULL in fkPELT")}
  n<-length(data)
  f<-numeric(length = n+1)
  f[1]<- -3*log(300) # 3*log(300) is the penalty constant tha we use
                     #in the PELT algorithm. Inspired by BIC criterion but
                     #modified
  cp<-rep(list(numeric()),n+1)
  R<-rep(list(numeric()),n)
  R[[1]]<-0


  for (t in 1:n){
    m<-length(R[[t]])
    neglog<-numeric(m)
    for (r in 1:m){
      neglog[r]<-fksplinecost(data[(R[[t]][r]+1):t],
                              knots=knots,index1=R[[t]][r]+1,index2=t)
    }
    stat<-numeric(m)
    for (r in 1:m){
      # Calculate the global cost with each possible change points in R[[t]]
      tot<-f[R[[t]][r]+1]+3*log(300)+neglog[r]
      stat[r]<-tot

    }
    f[t+1]<-min(stat)
    t1<-R[[t]][which.min(stat)]
    cp[[t+1]]<-c(cp[[t1+1]],t1)
    for (r in 1:m){

      tot<-f[R[[t]][r]+1]+log(300)+neglog[r]
      #Here we have the pruning which is amplified by this "log(300)"
      #, this makes the method a lot faster but a little bit less accurate
      if (tot<=f[t+1]& t<n){

        R[[t+1]]<-c(R[[t+1]],R[[t]][r])
      }
    }
    if (t<n&t>39){
      R[[t+1]]<-c(R[[t+1]],t-19)}
  }

  cp[[n+1]]<-cp[[n+1]][-1]
  return(cp[[n+1]]) #Return the set of change points
}


normalize.var<-function(data,endpoints){
  datanew<-changetometa(as.data.frame(data),endpoints)
  changes<-length(endpoints)
  ## Put in form that we can know which segment/plate the data is in
  colnames(datanew)<-"orig"
  if (changes>=1){
    ## if no change no normalization
    homo2<-homogen.var(datanew) # Test if the different segments
    # have homogeneity of variance
    if (!homo2){

      plates <-
        unique(unlist(lapply(strsplit(
          rownames(datanew), "_"
        ), function(x) {
          return(x[2])
        })))

      met.plates.sd <- c()

      for (j in plates) {

        grep.plates <-
          grep(paste("plate", "_", j, "_", sep = ""), row.names(datanew))
        sd.plate    <-  sd(datanew[grep.plates,], na.rm = TRUE)



        if(!is.na(sd.plate) & sd.plate == 0) {
          sd.plate <- NA
        }

        met.plates.sd <-
          c(met.plates.sd, rep(sd.plate, length(grep.plates)))
      }

      # Normalize
      datanew[["norm"]] <- datanew[["orig"]] / met.plates.sd
      if (sum(is.na(datanew[["norm"]]))>0){

        dataret<-datanew[["orig"]]

      }

      dataret<-datanew[["norm"]]
    }
  }

  dataret<-datanew[["orig"]]

  return(dataret)
}




